lib/foreignNodeId.ts

```diff
@@ -0,0 +1,61 @@
+import { DirectiveLocation, GraphQLNonNull, GraphQLString } from 'graphql';
+import { ValidationError } from 'apollo-server-errors';
+
+import ValidateDirectiveVisitor, {
+  ValidateFunction,
+} from './ValidateDirectiveVisitor';
+
+export type ToNodeId = (id: string) => { typename: string; id: string };
+
+export type ForeignNodeIdContext<TContext extends object = object> = {
+  fromNodeId: ToNodeId;
+};
+
+export type Args = {
+  typename: string;
+};
+
+export default class ForeignNodeIdDirective<
+  TContext extends ForeignNodeIdContext
+> extends ValidateDirectiveVisitor<Args, ForeignNodeIdContext> {
+  public getValidationForArgs():
+    | ValidateFunction<ForeignNodeIdContext>
+    | undefined {
+    return (value: unknown, _, __, { fromNodeId }): string => {
+      if (typeof value !== 'string') {
+        throw new ValidationError(
+          `${this.name} directive only works on strings`,
+        );
+      }
+      const { id, typename } = fromNodeId(value);
+      if (typename !== this.args.typename) {
```

Comment:
you can save `this.args.typename` before you `return (value:...)`, maybe this is the warning you were getting?

I'd also cache the error messages that you throw, see the other directives (createValidate).